### ISSUE TITLE
do not always import matplotlib

### ISSUE THREAD

**Issue by howthebodyworks** [librosa/librosa#343]
librosa always imports `matplotlib`, even if you are doing pure analysis and have no interest in graphing.

This is inconvenient for batch soundfile analysis, as matplotlib has a large memory footprint, and an even larger file handle footprint. Moreover, if you run matplotlib in certain environments (I'm using tornado to analyse soundfiles on request) it seems to leak file handles.

Whilst the file handle leak is not, AFAICT, librosa's problem per se, I've run into this issue in enough circumstances to feel that matplotlib should be considered something to be imported only when strictly required, and the memory problem alone is annoying when processing thousands of files in a multiprocess batch.

At first glance it seems that avoiding this could be as simple as only importing `librosa.display` on demand as opposed to in the root `__init__.py`.

Would it be worth me filing a pull request to do that?

**Comment by howthebodyworks** [comment]
related: #338

**Comment by howthebodyworks** [comment]
And while I'm here, let me mention also briefly my opinion that librosa is awesome.

**Comment by bmcfee** [comment]
This is redundant with #281 #314 #258 .

For the next release, we're not going to drop the import because it would break API.  I may revisit this for 0.5.0 since enough people have complained about it.

In the meantime, if it's really causing trouble, you could always suppress the import by mocking the module before loading librosa.  For example:

```python
import mock
import sys
sys.modules.update((mod_name, mock.Mock()) for mod_name in ['matplotlib', 'matplotlib.pyplot'])

import librosa
...
```

**Comment by TimSchmeier** [comment]
FYI: Instead of using the above hack this solved a number of librosa/matplotlib import issues for me

```
import matplotlib
matplotlib.use('Agg')
```

**Comment by carlthome** [comment]
I think it's bad practice for a package to import as needed. It adds unnecessary complexity and unexpected behavior. A package's memory requirements should not fluctuate depending on what functions you import. I expect a bunch of future issues where people wonder why they get OutOfMemory exceptions just because they called librosa.display.waveplot or specshow. If there's a strong use case for not having visualizations builtin then librosa should be split into two completely separate packages. One for analysis and one for visualizations, sharing a mutual interface.

**Comment by bmcfee** [comment]
> I think it's bad practice for a package to import as needed.

Agreed.  Dynamic submodule import is different though, and is standard practice in python (see, eg, sklearn).



> . If there's a strong use case for not having visualizations builtin then librosa should be split into two completely separate packages. One for analysis and one for visualizations, sharing a mutual interface.

I don't think there's a strong case for splitting this functionality.  Much of librosa's intended use is for research and prototyping, and visualizing output is critical there.  The display module really only provides two main functions (and a couple of helpers) to make the common use-cases easy, and I don't think that's enough to warrant a separate package.

**Comment by carlthome** [comment]
> Dynamic submodule import is different though, and is standard practice in python (see, eg, sklearn).
Cool! But typically that's a job for the the language's compiler, no? To optimize away unused imports, I mean. I suppose Python works differently because it's interpreted? I'm not particularly familiar with it, as you can probably tell, but I assumed the runtime would strip away unused code paths (but maybe not, considering that's impossible with a REPL and ipython, etc. seem very popular).

>I don't think there's a strong case for splitting this functionality.
Me neither, actually. Maybe in the future though? :smile:

**Comment by bmcfee** [comment]
> TL;DR; Just babbling, no need to reply or read.

I can't leave a thread dangling :scissors: 



> But typically that's a job for the the language's compiler, no? To optimize away unused imports, I mean. I suppose Python works differently because it's interpreted?

Exactly -- it's really no different from any other dynamic dependency management (eg, `dlopen()` or whatever incantations are necessary to link a `.dll` file at runtime).  If you really want dynamic functionality depending on user requirements, the logic for managing that ultimately bubbles up to the application level in one form or another.

Since I share your distaste for such things, I think I'd like to keep to the monolithic architecture we currently have.  However, since it's a common enough pain point, it might make sense to smooth over the rough edges of monkey-patching out mpl when it's not needed.  Then you could do something like
```python
import librosa.nodisplay
import librosa
# .. proceed as normal with matplotlib & friends mocked out
```
but that's a pretty brutal hack.



> I could certainly see myself using librosa on small/mid-sized production servers for apps and stuff in the future

:heart:

**Comment by danmackinlay** [comment]
Hello, it's @howthebodyworks writing from the office account.

There are a few of different issues here now.

One question is how much functionality a package provides; another is how much you are required to use.

There are different community standards around this for different languages, but the python one generally generally to _allow_ large packages (e.g. sklearn, scipy, django, twisted...) but to only _require_ you to use as much of each as you need (e.g. sklearn, scipy, django, twisted...) 

It's wonderful that librosa provides a lot of functionality; whether you get all or none of it, or allow some intermediate cases is a design decision. Sure, allowing only monolithic imports has costs *and* benefits, but I'd argue that the intuition of most python developers is built around granular imports, and so the theoretical benefits of making granular imports difficult will be cancelled out by the confused pythonistas complaining about it.

As @carlthome says:

> A package's memory requirements should not fluctuate heavily depending on what modules you import. I expect a bunch of future issues where people wonder why they get OutOfMemory exceptions just because they called librosa.display.waveplot or specshow.

Yes, people will absolutely wonder about this, but IMO it's not `librosa`'s problem. Our question is not which future has _no_ people wondering about computers being abstruse; there are no such futures. Rather, it's  which future has the _least_ people wondering about said abstruseness. 

So I'll mention another possible reason that the future I suggest is an abstrusness-wonder-minimising-future. ;-)

Another issue is _when_ we let memory run out. I absolutely agree that randomly importing chunks of code without user intervention at arbitrary times is bad, but I don't think that's what's being suggested here. Rather we are discussing importing chunks of code explicitly at user request, where users ar the developers using the code. If the developers are writing a nice GUI app for others they can import everything at front; If they are developing some crazy server farm audio analysis, they can import the non-GUI-parts.

Whilst it would be wonderful if we knew before launching any python instance how much memory it will request and when, this is not in general remotely possible; especially with exploratory data analysis via a REPL or jupyter, or any of these use-cases that make dynamic languages such as python powerful. Rather, everyone who uses these has to learn that OOM errors are always possible at `import` time, and manage their own risks. Relatedly, we can reduce the risk of this by hoping that package developers are a parsimonious as possible with importing dependencies - although that is more an argument _against_ importing `matplotlib` by default.

Moreover, even if we only allow the user to make every import they might ever need, right at the start of their python session, we still can't avoid OOM-errors occuring later. What if they load up an 8Gb sound file then take the STFT?

This is the point where someone usually mentions [the zen of python](https://www.python.org/dev/peps/pep-0020/):

>     Explicit is better than implicit.
>     Simple is better than complex.
>     Complex is better than complicated.

I'd argue that in that case, @bmcfee's potential solution
````
import librosa.nodisplay
import librosa
````
is not so much a brutal hack as "unpythonic"; I think python developers are more likely to expect explicit importing to explicitly import stuff, and be at ease with doing more imports if they want to import more. Also, it's uncomplicated. ;-)

````
import librosa
import librosa.display
````

this would change the number of import statements in my code, at least, by something like 10%, and reduce the number of out-of-file-handle errors by 100%. YMMV.

**Comment by bmcfee** [comment]
@danmackinlay @howthebodyworks I agree with all of that.

OTOH, if you're losing file handles by importing matplotlib, I suggest reporting that to matplotlib and not librosa. :grin:

**Comment by danmackinlay** [comment]
Yes, I haven't mentioned the details here because it's one of those irritating things that  lies at the intersection of several projects. But for other weary travellers, there  the issue is that [gc does not always close files](https://github.com/matplotlib/matplotlib/commit/4cb2aaa87b0aa9efe7178d7510d3d091ef5206df) at the best of times, and that the quirky matplotlib backend opens lots of files for reasons to do with fonts. The problem seems to arise especially often if you are using something that does live code reloading, such as [a web development server](http://www.tornadoweb.org/en/stable/autoreload.html).

There are a couple of long-delayed matplotlib backend changes which will radically change the font handling when they land, although they are many months in review already:
https://github.com/matplotlib/matplotlib/pull/5414
https://github.com/matplotlib/matplotlib/pull/4143

But in the mean time @TimSchmeier 's tip actually does ease the issue even though without the MPL backend; the Agg Matplotlib backend can run for a lot more code iterations (at least on my python) before running out of file handles. I don't know why.

**Comment by jorenvs** [comment]
We use librosa for signal processing only and use our own visualization tools when needed, but in production, where no visualization in python is needed, we really don't want to install and import matplotlib. It gives an error when you don't change the default backend, it slows down execution and makes deploys more flaky and requires us to install matplotlib's system requirements. All that for a library that's not being used in the first place.

**Comment by bmcfee** [comment]
Closed by merging #402
